distributed/scheduler.py

```diff
@@ -4237,104 +4238,143 @@ async def add_nanny(self) -> dict[str, Any]:
-        if len(tasks) > 1:
+        requested_keys = set(keys)
+        del keys
+        if len(dsk) > 1:
             self.log_event(
-                ["all", client], {"action": "update_graph", "count": len(tasks)}
+                ["all", client], {"action": "update_graph", "count": len(dsk)}
             )
+        self._pop_known_tasks(dsk, dependencies)
 
-        # Remove aliases
-        for k in list(tasks):
-            if tasks[k] is k:
-                del tasks[k]
+        if lost_keys := self._pop_lost_tasks(dsk, dependencies, requested_keys):
+            self.report({"op": "cancelled-keys", "keys": lost_keys}, client=client)
+            self.client_releases_keys(
+                keys=lost_keys, client=client, stimulus_id=stimulus_id
+            )
+        runnable, touched_tasks, new_tasks = self._generate_taskstates(
+            keys=requested_keys, dsk=dsk, dependencies=dependencies, code=code
+        )
 
-        dependencies = dependencies or {}
+        final_annotations = self._parse_and_apply_annotations(
+            tasks=new_tasks,
+            annotations=annotations,
+            layer_annotations=layer_annotations,
+            pre_stringified_keys=pre_stringified_keys,
+        )
+
+        self._set_priorities(
+            internal_priority=internal_priority,
+            submitting_task=submitting_task,
+            user_priority=user_priority,
+            fifo_timeout=fifo_timeout,
+            start=start,
+            dsk=dsk,
+            tasks=runnable,
+            dependencies=dependencies,
+        )
 
+        self.client_desires_keys(keys=requested_keys, client=client)
+
+        # Add actors
+        if actors is True:
+            actors = list(requested_keys)
+        for actor in actors or []:
+            ts = self.tasks[actor]
+            ts.actor = True
+
+        # Compute recommendations
+        recommendations: Recs = {}
+        priority = dict()
+        for ts in sorted(
+            runnable,
+            key=operator.attrgetter("priority"),
+            reverse=True,
+        ):
+            assert ts.priority  # mypy
+            priority[ts.key] = ts.priority
+            assert ts.run_spec
+            if ts.state == "released":
+                recommendations[ts.key] = "waiting"
+
+        for ts in runnable:
+            for dts in ts.dependencies:
+                if dts.exception_blame:
+                    ts.exception_blame = dts.exception_blame
+                    recommendations[ts.key] = "erred"
+                    break
+
+        for plugin in list(self.plugins.values()):
+            try:
+                plugin.update_graph(
+                    self,
+                    client=client,
+                    tasks=[ts.key for ts in touched_tasks],
+                    keys=requested_keys,
+                    dependencies=dependencies,
+                    annotations=final_annotations,
+                    priority=priority,
+                )
```

Comment:
The plugin API is something I'm not entirely sure what to do with. I wrote it now such that our tests pass but the API did change, I'm not passing all arguments as before and unless we specifiy more explicitly what the arguments are, I believe we should pass them at all

The docs don't say anything https://distributed.dask.org/en/stable/plugins.html#distributed.diagnostics.plugin.SchedulerPlugin.update_graph

For instance, I was surprised to see that `tasks` are actually keys...

Reply:
> One user in particular I care about is @ntabris who might use this for analytics.

I checked. From what I can tell, the plugin uses only the transition hook but not the update_graph hook

Reply:
No concerns from me, like @fjetter says we aren't using graph updates.